### PR TITLE
Add desktop projects and build commands (experimental)

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

Adds experimental build commands for desktop targets which passes the flutter root.

## Related Issues

https://github.com/flutter/flutter/issues/30706

## Tests

I added the following tests: lots

*Replace this with a list of the tests that you added as part of this PR. A change in behaviour with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flutter developers to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (Please read [Handling breaking changes]). *Replace this with a link to the e-mail where you asked for input on this proposed change.*
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/build_linux.dart [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+/// A command to build a linux desktop target through a build shell script.
+class BuildLinuxCommand extends BuildSubCommand {
+  @override
+  final String name = 'linux';
+
+  @override
+  bool isExperimental = true;
+
+  @override
+  String get description => 'build the linux desktop target (Experimental).';
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_linux.dart]: s/linux/Linux/

File: packages/flutter_tools/lib/src/commands/build_linux.dart [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final FlutterProject flutterProject = await FlutterProject.current();
+    if (!flutterProject.linux.existsSync()) {
+      throwToolExit('No linux desktop project configured.');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_linux.dart]: s/linux/Linux/

File: packages/flutter_tools/lib/src/commands/build_macos.dart [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+/// A command to build a macos desktop target through a build shell script.
+class BuildMacosCommand extends BuildSubCommand {
+  @override
+  final String name = 'macos';
+
+  @override
+  bool isExperimental = true;
+
+  @override
+  String get description => 'build the MacOS desktop target (Experimental).';
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_macos.dart]: macOS

File: packages/flutter_tools/lib/src/commands/build_macos.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final FlutterProject flutterProject = await FlutterProject.current();
+    if (!flutterProject.macos.existsSync()) {
+      throwToolExit('No MacOS desktop project configured.');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_macos.dart]: macOS

File: packages/flutter_tools/lib/src/commands/build_windows.dart [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+/// A command to build a windows desktop target through a build shell script.
+class BuildWindowsCommand extends BuildSubCommand {
+  @override
+  final String name = 'windows';
+
+  @override
+  bool isExperimental = true;
+
+  @override
+  String get description => 'build the desktop windows target (Experimental).';
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_windows.dart]: Windows

File: packages/flutter_tools/lib/src/commands/build_windows.dart [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+  @override
+  Future<FlutterCommandResult> runCommand() async {
+    final FlutterProject flutterProject = await FlutterProject.current();
+    if (!flutterProject.windows.existsSync()) {
+      throwToolExit('No windows desktop project configured.');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/commands/build_windows.dart]: Windows

File: packages/flutter_tools/lib/src/project.dart [hunk 7]
```diff
@@ -518,3 +527,33 @@ Match _firstMatchInFile(File file, RegExp regExp) {
...
 }
+
+class MacOSProject {
+  MacOSProject._(this.project);
+
+  final FlutterProject project;
+
+  bool existsSync() => project.directory.childDirectory('macos').existsSync();
+
+  File get buildScript => project.directory.childDirectory('macos').childFile('build.sh');
```
Comment (stuartmorgan) [inline packages/flutter_tools/lib/src/project.dart]: It is worth documenting these to explain that this is a temporary shim?

**Review by stuartmorgan-g — APPROVED** [review]

**Pull request merged**